src/metabase/pulse.clj

```diff
@@ -44,40 +44,59 @@
-  [{:keys [id name] :as pulse} results recipients]
+(defn- create-email-notification [{:keys [id name] :as pulse} results recipients]
   (log/debug (format "Sending Pulse (%d: %s) via Channel :email" id name))
   (let [email-subject    (str "Pulse: " name)
         email-recipients (filterv u/is-email? (map :email recipients))
         timezone         (-> results first :card defaulted-timezone)]
-    (email/send-message!
-      :subject      email-subject
-      :recipients   email-recipients
-      :message-type :attachments
-      :message      (messages/render-pulse-email timezone pulse results))))
+    {:subject      email-subject
+     :recipients   email-recipients
+     :message-type :attachments
+     :message      (messages/render-pulse-email timezone pulse results)}))
 
-(defn create-and-upload-slack-attachments!
-  "Create an attachment in Slack for a given Card by rendering its result into an image and uploading it."
+(defn- send-email-pulse!
+  "Send a `Pulse` email given a list of card results to render and a list of recipients to send to."
+  [{:keys [subject recipients message-type message]}]
+  (email/send-message!
+    :subject      subject
+    :recipients   recipients
+    :message-type message-type
+    :message      message))
+
+(defn create-slack-attachment-data
+  "Returns a seq of slack attachment data structures, used in `create-and-upload-slack-attachments!`"
   [card-results]
   (let [{channel-id :id} (slack/files-channel)]
-    (doall (for [{{card-id :id, card-name :name, :as card} :card, result :result} card-results]
-             (let [image-byte-array (render/render-pulse-card-to-png (defaulted-timezone card) card result)
-                   slack-file-url   (slack/upload-file! image-byte-array "image.png" channel-id)]
-               {:title      card-name
-                :title_link (urls/card-url card-id)
-                :image_url  slack-file-url
-                :fallback   card-name})))))
+    (for [{{card-id :id, card-name :name, :as card} :card, result :result} card-results]
+      {:title      card-name
+       :attachment-bytes-thunk (fn [] (render/render-pulse-card-to-png (defaulted-timezone card) card result))
+       :title_link (urls/card-url card-id)
+       :attachment-name "image.png"
+       :channel-id channel-id
+       :fallback   card-name})))
+
+(defn- create-slack-notification [pulse results channel-id]
+  (log/debug (u/format-color 'cyan "Sending Pulse (%d: %s) via Slack" (:id pulse) (:name pulse)))
+  {:channel-id channel-id
+   :message (str "Pulse: " (:name pulse))
+   :attachments (create-slack-attachment-data results)})
+
+(defn create-and-upload-slack-attachments!
+  "Create an attachment in Slack for a given Card by rendering its result into an image and uploading it."
+  [attachments]
+  (doall
+   (for [{:keys [attachment-bytes-thunk attachment-name channel-id] :as attachment-data} attachments]
```

Comment:
why not just use `doseq` here instead of `doall` + `for`?

Reply:
The result of the `for` is still being used for sending the slack message later, so `doseq` wouldn't work. Not sure why the `doall` is necessary there, other than to maybe avoid uploading a partial message if the server was to crash in the middle of one of the uploads (i.e. laziness would upload a file at a time as the message was being sent, instead of uploading the files first, then sending the message with reference to those files).